src/MiddlewareListener.php

```diff
@@ -44,20 +44,38 @@ public function attach(EventManagerInterface $events, $priority = 1)
     public function onDispatch(MvcEvent $event)
     {
         if (null !== $event->getResult()) {
-            return;
+            return null;
         }
 
         $routeMatch = $event->getRouteMatch();
+
+        if (null === $routeMatch) {
+            throw new RuntimeException('No RouteMatch in event');
+        }
+
         $middleware = $routeMatch->getParam('middleware', false);
         if (false === $middleware) {
-            return;
+            return null;
         }
 
         $request        = $event->getRequest();
         $application    = $event->getApplication();
+
+        if (null === $request) {
+            throw new UnexpectedValueException('No Request in event');
+        }
+
+        if (! $application instanceof Application) {
```

Comment:
There is a method (`marshalInvalidMiddleware`) called some lines lower, where the signature require an `Application` instance. The interface would be enough but to avoid a BC break I can't change the signature.